trie/trie.go

```diff
@@ -1026,14 +1050,73 @@ func (t *Trie) resolveShadowNode(epoch types.StateEpoch, cur node, prefix []byte
 			}
 		}
 		return nil
-	case valueNode, hashNode:
+	case valueNode, hashNode, nil:
 		// just skip
 		return nil
 	default:
 		return errors.New("resolveShadowNode unsupported node type")
 	}
 }
 
+func (t *Trie) ShadowHash() (*common.Hash, error) {
+	if t.root == nil {
+		return nil, nil
+	}
+	h := newHasher(true)
+	defer returnHasherToPool(h)
+	return t.shadowHash(t.root, h, nil, t.root.getEpoch())
+}
+
+// shadowHash calculate node's shadow node hash, recalculate needn't a copy
+func (t *Trie) shadowHash(origin node, h *hasher, prefix []byte, epoch types.StateEpoch) (*common.Hash, error) {
+	switch n := origin.(type) {
+	case *shortNode:
+		var err error
+		if n.shadowNode.ShadowHash, err = t.shadowHash(n.Val, h, append(prefix, n.Key...), epoch); err != nil {
+			return nil, err
+		}
+		return h.shadowExtendNodeToHash(&n.shadowNode), nil
+	case *fullNode:
+		epochSelf := n.epoch
+		epochMap := n.shadowNode.EpochMap
+		hashList := make([]*common.Hash, 0, BranchNodeLength-1)
+		for i := byte(0); i < BranchNodeLength-1; i++ {
+			child := n.Children[i]
+			if child == nil {
+				continue
+			}
+			// skip expired node.
+			if epochSelf >= epochMap[i]+2 {
+				continue
+			}
+
+			subHash, err := t.shadowHash(child, h, append(prefix, i), epochMap[i])
+			if err != nil {
+				return nil, err
+			}
+			if subHash != nil {
+				hashList = append(hashList, subHash)
+			}
+		}
+		n.shadowNode.ShadowHash = h.shadowNodeHashListToHash(hashList)
+		return h.shadowBranchNodeToHash(&n.shadowNode), nil
+	case valueNode:
+		return nil, nil
+	case hashNode:
+		// resolve temporary, not add to trie
+		rn, err := t.resolveHash(n, prefix)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
why no need to add the resolved node to trie? 

I'm afraid that the calculation of shadow hash might be quite frequent, so the I/O cost might be higher?

Reply:
the calculation of shadow hash only once when trie commits, and only fullNode needs resolveHash to calculate its shadowHash, another reason is the later committer will commit all nodes to db, so the temporary node only using in the cache.

The trie node only resolves one time in shadowHash calculation, and is useless in later logic, so I think not adding to trie is ok.